### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ruancomelli**
# Description

_Explain briefly about the purpose & context of this PR_

# Screenshots

_If possible add a screenshot and/or gif_

# Example | Usage

_Please share examples and how to reproduce_

---

# Related Issues | Pr

_`close`|`resolve` <ISSUE> (ex. `resolve #20`)_
_link issues to close automatically_

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `multiple_mermaid_diagram_types`
Run ID: `multiple_mermaid_diagram_types/run_2024-09-30T19-14-26_v1-23-0-43-g5f7f3998a`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/timba-api/pull/12.
This pull request was cloned from `https://github.com/AlphaDev87/timba-api/pull/49`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
  # Sourcery core examples:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
  # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
  # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
  # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # A ton of new classes:
- https://github.com/sourcery-ai/core/pull/4618
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic
  # Brilliant existing examples:
- https://github.com/sourcery-ai/core/pull/4680
- https://github.com/sourcery-ai/core/pull/4684

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/374
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/375
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/376
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/5
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/13
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/14
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/14
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/13
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/4
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/23
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/24
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/7
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/13
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/14
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/7
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/7
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/13
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/14
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/7
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/7
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/7
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/6
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/14
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/15
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/15
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/14
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/7
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/6
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/11
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/12
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4579
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/377

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request implements significant changes to the project structure, database schema, and functionality. Key changes include refactoring the transactions system, improving error handling, adding new features like password reset and analytics, and enhancing security measures.

#### Updated ER diagram for Player and Payment entities

```mermaid
erDiagram
    PLAYER {
        string id
        int panel_id
        string username
        string password
        string email
        string first_name
        string last_name
        date date_of_birth
        string movile_number
        string country
        string balance_currency
        string status
        datetime created_at
        datetime updated_at
    }
    PAYMENT {
        string id
        string player_id
        float amount
        string bank_account
        string currency
        string status
        boolean dirty
        int alquimia_id
        datetime created_at
        datetime updated_at
    }
    PLAYER ||--o{ PAYMENT : "has many"
```

#### Updated class diagram for PlayerServices

```mermaid
classDiagram
    class PlayerServices {
        +create(player: PlayerRequest) Promise<PlainPlayerResponse>
        +login(credentials: Credentials, user_agent: string) Promise<LoginResponse>
        +resetPassword(user: Player, new_password: string) Promise<void>
        +update(player_id: string, request: PlayerUpdateRequest) Promise<PlainPlayerResponse>
        -createCasinoPlayer(username: string, password: string) Promise<number>
        -createLocalPlayer(panel_id: number, player: PlayerRequest) Promise<PlainPlayerResponse>
        -updatePlayerPassword(credentials: Credentials, id: number, email: string) Promise<Player>
        -emailPasswordResetConfirmation(user: Player)
    }
```

#### Updated class diagram for DepositServices

```mermaid
classDiagram
    class DepositServices {
        +create(player: PlainPlayerResponse, request: DepositRequest) Promise<DepositResult>
        +confirm(player: RoledPlayer, deposit_id: string, request: DepositRequest) Promise<DepositResult>
        +verify(deposit: Deposit) Promise<number | undefined>
        +showPending(player_id: string) Promise<Deposit[]>
        +pendingCoinTransfers() Promise<number>
        -alquimiaDepositLookup(tracking_number: string) Promise<AlqMovementResponse | undefined>
        -markAsPending(deposit: Deposit) Promise<Deposit & { Player: Player }>
        -markAsVerified(deposit: Deposit, amount: number) Promise<Deposit & { Player: Player }>
        -markAsConfirmed(deposit: Deposit) Promise<Deposit & { Player: Player }>
        -markAsCompleted(deposit: Deposit) Promise<Deposit & { Player: Player }>
    }
```

#### Updated class diagram for AgentServices

```mermaid
classDiagram
    class AgentServices {
        +login(credentials: Credentials) Promise<Tokens>
        +releasePayment(payment_id: string) Promise<Payment>
        +getBankAccount() Promise<AgentBankAccount>
        +freePendingCoinTransfers() Promise<Deposit[]>
        +getSupportNumbers() Promise<SupportResponse>
        +updateSupportNumbers(data: UserRootUpdatableProps) Promise<SupportResponse>
    }
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Refactored transactions system | <ul><li>Split transactions into separate deposits and payments components</li><li>Updated database schema for deposits and payments</li><li>Implemented new services for handling deposits and payments</li><li>Added rate limiting for payment releases</li></ul> | `src/components/deposits/services.ts`<br/>`src/components/payments/services.ts`<br/>`src/db/deposits.ts`<br/>`src/db/payments.ts`<br/>`src/routes/transactions/deposits.router.ts`<br/>`src/routes/transactions/payments.router.ts` |
| Improved error handling and logging | <ul><li>Implemented custom error classes</li><li>Updated error middleware to handle different types of errors</li><li>Added more detailed error logging</li><li>Standardized API response format</li></ul> | `src/middlewares/errorHandler.ts`<br/>`src/helpers/error/CustomError.ts`<br/>`src/helpers/error/AgentApiError.ts`<br/>`src/helpers/error/AlquimiaApiError.ts`<br/>`src/helpers/apiResponse.ts` |
| Added password reset functionality | <ul><li>Implemented password reset request and token generation</li><li>Created email service for sending password reset instructions</li><li>Added new routes and controllers for password reset flow</li></ul> | `src/components/new-password/services.ts`<br/>`src/components/new-password/controller.ts`<br/>`src/helpers/email/email.ts`<br/>`src/routes/auth.router.ts` |
| Implemented analytics system | <ul><li>Created analytics table in database</li><li>Added routes and controllers for creating and retrieving analytics data</li><li>Implemented analytics service</li></ul> | `prisma/migrations/20240605135628_create_analytics_table/migration.sql`<br/>`src/routes/analytics.router.ts`<br/>`src/components/analytics/controller.ts`<br/>`src/components/analytics/services.ts` |
| Enhanced security measures | <ul><li>Implemented rate limiting for password reset requests</li><li>Added validation for bank account numbers (CLABE)</li><li>Improved input validation across various endpoints</li></ul> | `src/middlewares/rate-limiters/pass-restore-request.ts`<br/>`src/components/bank-accounts/validators.ts`<br/>`src/components/players/validators.ts` |
| Refactored and improved player management | <ul><li>Updated player creation and update processes</li><li>Implemented pagination and search for player listing</li><li>Added ability for agents to reset player passwords</li></ul> | `src/components/players/services.ts`<br/>`src/components/players/controller.ts`<br/>`src/routes/players.router.ts` |
| Added WhatsApp integration | <ul><li>Implemented WhatsApp notification service</li><li>Added welcome message sending on player creation</li></ul> | `src/notification/whatsapp.ts`<br/>`src/components/players/services.ts` |
| Updated database schema and migrations | <ul><li>Added new tables for analytics and chat data</li><li>Updated existing tables with new fields and constraints</li><li>Removed deprecated tables and fields</li></ul> | `prisma/migrations/20240508181836_initial_migration/migration.sql`<br/>`prisma/migrations/20240514142314_create_chat_data_table/migration.sql`<br/>`prisma/migrations/20240605135628_create_analytics_table/migration.sql` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/routes/agent.router.ts [hunk 1]
```diff
@@ -67,5 +54,19 @@ agentRouter.post(
...
+  checkExact(),
+  throwIfBadRequest,
+  AgentController.updateSupportNumbers,
+);
+agentRouter.post(
+  "/reset-player-password",
+  validateResetPasswordRequest(),
+  checkExact(),
+  throwIfBadRequest,
+  AgentController.resetPlayerPassword,
```
Comment (SourceryAI) [inline src/routes/agent.router.ts:70]: **🚨 suggestion (security):** New password reset functionality needs security review

The addition of a password reset functionality for agents is a sensitive operation. Ensure that this endpoint is properly secured, audited, and only accessible to authorized agents. Consider implementing additional safeguards such as requiring secondary authentication for this operation.

File: src/db/payments.ts [hunk 2]
```diff
@@ -111,4 +141,30 @@ export class PaymentsDAO {
       prisma.$disconnect();
     }
   }
+
+  static async authorizeRelease(
```
Comment (SourceryAI) [inline src/db/payments.ts:145]: **suggestion:** Critical operation for payment release, needs careful review

The authorizeRelease method is a critical operation for payment processing. Ensure that all edge cases are handled, such as concurrent requests or system failures during the transaction. Consider adding more detailed logging or an audit trail for this operation.

File: src/db/players.ts [hunk 3]
```diff
@@ -1,26 +1,73 @@
...
+    try {
+      return prisma.player.count();
+    } catch (e) {
+      throw e;
+    } finally {
+      prisma.$disconnect();
+    }
+  }
+
+  static _getAll = async (
```
Comment (SourceryAI) [inline src/db/players.ts:30]: **suggestion (performance):** New pagination and search methods need performance optimization

The addition of pagination and search functionality is good for scalability. However, ensure that these queries are optimized for performance, especially for large datasets. Consider adding indexes on frequently searched fields and test the performance with a large number of records.

File: docker-compose.yml [hunk 4]
```diff
@@ -1,21 +1,20 @@
...
 # AlphaDev casinomex database
 services:
-  casinomex_db:
-    image: mysql
+  db:
+    image: mariadb
     restart: unless-stopped
     environment:
-      MYSQL_ROOT_PASSWORD: timba2024
-      MYSQL_DATABASE: timba-data-base
```
Comment (SourceryAI) [inline docker-compose.yml:10]: **🚨 suggestion (security):** Update database environment variables

Ensure that the new database credentials are properly secured and not exposed in version control. Consider using environment variables or secrets management for sensitive information.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,137 @@
+## [1.0.0] - 2024-05-29
```
Comment (SourceryAI) [inline CHANGELOG.md:1]: **suggestion (documentation):** Consider standardizing the date format throughout the changelog

Some entries use YYYY-MM-DD format while others use YYYY-DD-MM. Consistency would improve readability.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,137 @@
...
+- Split GET `/agent/deposits/:id?` into GET `/transactions/deposit/:id` and GET `/transactions/deposit`.
+- Move POST `/agent/deposits/:id` into `/transactions/deposit/:id`.
+- Return `{ deposits, totalDeposits }` from GET `/transactions/deposit`.
+- Move GET `/agent/payments` to GET `/transactions/payment`
+- Return `{ payments, totalPayments }` from GET `/transactions/payment`.
+
+## [1.0.0] - 2024-05-27
+
+### Changed
+- Improve `movile_number` validator: must be a numerical string of up to 20 characters
```
Comment (SourceryAI) [inline CHANGELOG.md:21]: **issue (documentation):** Correct typo in 'movile_number'

The correct spelling is 'mobile_number'.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟡 **Documentation**: 2 issues found
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/fd8f5d16-1e06-4b41-a3aa-f9c2126ddab9?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Pull request closed without merging**